File: wrappers/java/phe/src/test/java/com/virgilsecurity/crypto/phe/PheClientTest.java
/*
* Copyright (C) 2015-2019 Virgil Security, Inc.
*
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are
* met:
*
* (1) Redistributions of source code must retain the above copyright
* notice, this list of conditions and the following disclaimer.
*
* (2) Redistributions in binary form must reproduce the above copyright
* notice, this list of conditions and the following disclaimer in
* the documentation and/or other materials provided with the
* distribution.
*
* (3) Neither the name of the copyright holder nor the names of its
* contributors may be used to endorse or promote products derived from
* this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
* IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
* INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
* STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
* IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*
* Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
*/

package com.virgilsecurity.crypto.phe;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.nio.charset.StandardCharsets;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

public class PheClientTest {

	private PheClient client;
	private PheServer server;

	@Before
	public void setup() {
		this.client = new PheClient();
		this.client.setupDefaults();
		assertTrue("C context should be set", this.client.cCtx > 0);

		this.server = new PheServer();
		this.server.setupDefaults();
		assertTrue("C context should be set", this.server.cCtx > 0);
	}

	@After
	public void teardown() {
		this.client.close();
		this.server.close();
	}

	@Test
	public void testFullFlowRandomCorrectPwdShouldSucceed() throws PheException {
		byte[] password = "password".getBytes(StandardCharsets.UTF_8);

		PheServerGenerateServerKeyPairResult serverKeyPair = server.generateServerKeyPair();
		assertNotNull(serverKeyPair);

		byte[] serverPrivateKey = serverKeyPair.getServerPrivateKey();
		assertNotNull(serverPrivateKey);
		assertEquals(32, serverPrivateKey.length);

		byte[] serverPublicKey = serverKeyPair.getServerPublicKey();
		assertNotNull(serverPublicKey);
		assertEquals(65, serverPublicKey.length);

		byte[] clientPrivateKey = this.client.generateClientPrivateKey(); // {privateKey}
		assertNotNull(clientPrivateKey);

		this.client.setKeys(clientPrivateKey, serverPublicKey); // void

		byte[] serverEnrollment = this.server.getEnrollment(serverPrivateKey, serverPublicKey);
		assertNotNull(serverEnrollment);
		assertTrue(serverEnrollment.length > 0);

		PheClientEnrollAccountResult clientEnrollAccount = this.client.enrollAccount(serverEnrollment, password);
		assertNotNull(clientEnrollAccount);

		byte[] clientEnrollmentRecord = clientEnrollAccount.getEnrollmentRecord();
		assertNotNull(clientEnrollmentRecord);

		byte[] clientAccountKey = clientEnrollAccount.getAccountKey();
		assertNotNull(clientAccountKey);
		assertEquals(32, clientAccountKey.length);

		byte[] clientCreateVerifyPasswordRequest = this.client.createVerifyPasswordRequest(password,
				clientEnrollmentRecord);
		assertNotNull(clientCreateVerifyPasswordRequest);
		assertTrue(clientCreateVerifyPasswordRequest.length > 0);

		byte[] serverVerifyPassword = this.server.verifyPassword(serverPrivateKey, serverPublicKey,
				clientCreateVerifyPasswordRequest);
		assertNotNull(serverVerifyPassword);

		byte[] clientCheckResponseAndDecrypt = this.client.checkResponseAndDecrypt(password, clientEnrollmentRecord,
				serverVerifyPassword);
		assertNotNull(clientCheckResponseAndDecrypt);
		assertEquals(32, clientCheckResponseAndDecrypt.length);
		assertArrayEquals(clientAccountKey, clientCheckResponseAndDecrypt);
	}

	@Test
	public void testRotationRandomRotationServerPublicKeysMatch() throws PheException {
		PheServerGenerateServerKeyPairResult serverKeyPair = this.server.generateServerKeyPair();
		assertNotNull(serverKeyPair);

		byte[] serverPrivateKey = serverKeyPair.getServerPrivateKey();
		assertNotNull(serverPrivateKey);

		byte[] serverPublicKey = serverKeyPair.getServerPublicKey();
		assertNotNull(serverPublicKey);

		PheServerRotateKeysResult serverRotateKeys = this.server.rotateKeys(serverPrivateKey);
		assertNotNull(serverRotateKeys);

		byte[] serverRotatedPrivateKey = serverRotateKeys.getNewServerPrivateKey();
		assertNotNull(serverRotatedPrivateKey);
		assertEquals(32, serverRotatedPrivateKey.length);

		byte[] serverRotatedPublicKey = serverRotateKeys.getNewServerPublicKey();
		assertNotNull(serverRotatedPublicKey);
		assertEquals(65, serverRotatedPublicKey.length);

		byte[] serverUpdateToken = serverRotateKeys.getUpdateToken();
		assertNotNull(serverUpdateToken);
		assertTrue(serverUpdateToken.length > 0);

		byte[] clientPrivateKey = this.client.generateClientPrivateKey();
		assertNotNull(clientPrivateKey);
		assertEquals(32, clientPrivateKey.length);

		this.client.setKeys(clientPrivateKey, serverRotatedPublicKey);

		PheClientRotateKeysResult clientRotateKeys = this.client.rotateKeys(serverUpdateToken);
		assertNotNull(clientRotateKeys);

		byte[] clientNewPrivateKey = clientRotateKeys.getNewClientPrivateKey();
		assertNotNull(clientNewPrivateKey);
		assertEquals(32, clientNewPrivateKey.length);

		byte[] serverNewPublicKey = clientRotateKeys.getNewServerPublicKey();
		assertNotNull(serverNewPublicKey);
		assertEquals(65, serverNewPublicKey.length);

		assertEquals(serverPublicKey.length, serverNewPublicKey.length);
		assertEquals(clientPrivateKey.length, clientNewPrivateKey.length);
	}

}

File: wrappers/go/phe/phe_client_test.go
//  Copyright (C) 2015-2020 Virgil Security, Inc.
//
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//      (1) Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//      (2) Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in
//      the documentation and/or other materials provided with the
//      distribution.
//
//      (3) Neither the name of the copyright holder nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
//  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
//  INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
//  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
//  HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
//  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
//  IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
//  Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>

package phe

import (
	"github.com/stretchr/testify/require"
	"testing"
)

func TestNewPheClient(t *testing.T) {
	client := NewPheClient()
    require.NotNil(t, client)

	client.Delete()
}

func TestFullFlowRandomCorrectPwdShouldSucceed(t *testing.T) {
	password := []byte("password")

	client := newPheClient()
	defer client.Delete()

	server := newPheServer()
	defer server.Delete()

	serverPrivateKey, serverPublicKey, err := server.GenerateServerKeyPair()
	require.Nil(t, err)
	require.Equal(t, 32, len(serverPrivateKey))
	require.Equal(t, 65, len(serverPublicKey))

	clientPrivateKey, err := client.GenerateClientPrivateKey()
	require.Nil(t, err)
	require.NotNil(t, clientPrivateKey)

	err = client.SetKeys(clientPrivateKey, serverPublicKey)
	require.Nil(t, err)

	serverEnrollment, err := server.GetEnrollment(serverPrivateKey, serverPublicKey)
	require.Nil(t, err)
	require.NotNil(t, serverEnrollment)
	require.True(t, len(serverEnrollment) > 0)

	clientEnrollmentRecord, clientAccountKey, err := client.EnrollAccount(serverEnrollment, password)
	require.Nil(t, err);
	require.NotNil(t, clientEnrollmentRecord)
	require.NotNil(t, clientAccountKey)
	require.Equal(t, 32, len(clientAccountKey))

	clientCreateVerifyPasswordRequest, err := client.CreateVerifyPasswordRequest(password, clientEnrollmentRecord)
	require.Nil(t, err)
	require.NotNil(t, clientCreateVerifyPasswordRequest)
	require.True(t, len(clientCreateVerifyPasswordRequest) > 0)

	serverVerifyPassword, err := server.VerifyPassword(serverPrivateKey, serverPublicKey, clientCreateVerifyPasswordRequest)
	require.Nil(t, err)
	require.NotNil(t, serverVerifyPassword)

	clientCheckResponseAndDecrypt, err := client.CheckResponseAndDecrypt(password, clientEnrollmentRecord, serverVerifyPassword)
	require.Nil(t, err)
	require.NotNil(t, clientCheckResponseAndDecrypt)
	require.Equal(t, 32, len(clientCheckResponseAndDecrypt))
	require.Equal(t, clientAccountKey, clientCheckResponseAndDecrypt)
}

func TestRotationRandomRotationServerPublicKeysMatch(t *testing.T) {
	client := newPheClient()
	defer client.Delete()

	server := newPheServer()
	defer server.Delete()

	serverPrivateKey, serverPublicKey, err := server.GenerateServerKeyPair()
	require.Nil(t, err)
	require.NotNil(t, serverPrivateKey)
	require.NotNil(t, serverPublicKey)

	serverRotatedPrivateKey, serverRotatedPublicKey, serverUpdateToken, err := server.RotateKeys(serverPrivateKey)
	require.Nil(t, err)

	require.NotNil(t, serverRotatedPrivateKey)
	require.Equal(t,32, len(serverRotatedPrivateKey))

	require.NotNil(t, serverRotatedPublicKey)
	require.Equal(t, 65, len(serverRotatedPublicKey))

	require.NotNil(t, serverUpdateToken)
	require.True(t, len(serverUpdateToken) > 0)

	clientPrivateKey, err := client.GenerateClientPrivateKey()
	require.Nil(t, err)
	require.NotNil(t, clientPrivateKey)
	require.Equal(t, 32, len(clientPrivateKey))

	err = client.SetKeys(clientPrivateKey, serverRotatedPublicKey)
	require.Nil(t, err)

	clientNewPrivateKey, serverNewPublicKey, err := client.RotateKeys(serverUpdateToken)
	require.Nil(t, err)

	require.NotNil(t, clientNewPrivateKey)
	require.Equal(t, 32, len(clientNewPrivateKey))

	require.NotNil(t, serverNewPublicKey)
	require.Equal(t,65, len(serverNewPublicKey))

	require.Equal(t, len(serverPublicKey), len(serverNewPublicKey))
	require.Equal(t, len(clientPrivateKey), len(clientNewPrivateKey))
}

func newPheClient() *PheClient {
	client := NewPheClient()
	_ = client.SetupDefaults()

	return client
}